serdes-native/src/main/java/io/ballerina/stdlib/serdes/Deserializer.java

```diff
@@ -0,0 +1,301 @@
+        }
+        return object;
+    }
+
+    private static DynamicMessage generateDynamicMessageFromBytes(Descriptor schema, BArray encodedMessage)
+                                                                                throws InvalidProtocolBufferException {
+        return DynamicMessage.parseFrom(schema, encodedMessage.getBytes());
+    }
+
+    private static Object dynamicMessageToBallerinaType(DynamicMessage dynamicMessage, BTypedesc typedesc,
+                                                        Descriptor schema) {
+        Type type = typedesc.getDescribingType();
+
+        if (type.getTag() <= TypeTags.BOOLEAN_TAG) {
+            FieldDescriptor fieldDescriptor = schema.findFieldByName(ATOMIC_FIELD_NAME);
+            return getBallerinaPrimitiveValueFromMessage(dynamicMessage.getField(fieldDescriptor));
+        } else if (type.getTag() == TypeTags.UNION_TAG) {
+            FieldDescriptor fieldDescriptor = schema.findFieldByName(ATOMIC_FIELD_NAME);
+            DynamicMessage dynamicMessageForUnion = (DynamicMessage) dynamicMessage.getField(fieldDescriptor);
+            return getBallerinaUnionTypeValueFromMessage(dynamicMessageForUnion, type, schema);
+        } else if (type.getTag() == TypeTags.ARRAY_TAG) {
+            ArrayType arrayType = (ArrayType) type;
+            Type elementType = arrayType.getElementType();
+            FieldDescriptor fieldDescriptor = schema.findFieldByName(ARRAY_FIELD_NAME);
+            schema = fieldDescriptor.getContainingType();
+            return getBallerinaArrayValueFromMessage(dynamicMessage.getField(fieldDescriptor), elementType, schema, 1);
+        } else if (type.getTag() == TypeTags.RECORD_TYPE_TAG) {
+            Map<String, Object> mapObject = getBallerinaRecordValueFromMessage(dynamicMessage, type, schema);
+            return ValueCreator.createRecordValue(type.getPackage(), type.getName(), mapObject);
+        } else {
+            throw createSerdesError(UNSUPPORTED_DATA_TYPE + type.getName(), SERDES_ERROR);
+        }
+    }
+
+    private static Object getBallerinaPrimitiveValueFromMessage(Object value) {
+        if (value instanceof String) {
+            return StringUtils.fromString((String) value);
+        }
+        return value;
+    }
+
+    private static Object getBallerinaArrayValueFromMessage(Object value, Type type, Descriptor schema,
+                                                            int unionFieldIdentifier) {
+        if (value instanceof ByteString) {
+            ByteString byteString = (ByteString) value;
+            return ValueCreator.createArrayValue(byteString.toByteArray());
+        } else {
+            Collection collection = (Collection) value;
+            BArray bArray = ValueCreator.createArrayValue(TypeCreator.createArrayType(type));
+            for (Object element : collection) {
+                if (type.getTag() == TypeTags.STRING_TAG) {
+                    bArray.append(StringUtils.fromString((String) element));
```

Comment:
Let's cast the `element` to `BString` and use `stringValue()` method to get the value here.

Reply:
`element` is a Java String value. It creates a `ClassCastException` when cast to `BString`